api/handlers/exporter.go

```diff
@@ -95,3 +233,177 @@ func transformToParticipantResponse(role spectypes.BeaconRole, entry qbftstorage
+
+func (e *Exporter) ValidatorTraces(w http.ResponseWriter, r *http.Request) error {
+	var request struct {
+		From    uint64          `json:"from"`
+		To      uint64          `json:"to"`
+		Roles   api.RoleSlice   `json:"roles"`
+		PubKeys api.HexSlice    `json:"pubkeys"`
+		Indices api.Uint64Slice `json:"indices"`
+	}
+
+	if err := api.Bind(r, &request); err != nil {
+		return api.BadRequestError(err)
+	}
+
+	if request.From > request.To {
+		return api.BadRequestError(fmt.Errorf("'from' must be less than or equal to 'to'"))
+	}
+
+	if len(request.Roles) == 0 {
+		return api.BadRequestError(fmt.Errorf("at least one role is required"))
+	}
+
+	if len(request.PubKeys) == 0 && len(request.Indices) == 0 {
+		return api.BadRequestError(errors.New("either pubkeys or indices is required"))
+	}
+
+	var pubkeys []spectypes.ValidatorPK
+
+	for _, index := range request.Indices {
+		share, found := e.Validators.ValidatorByIndex(phase0.ValidatorIndex(index))
+		if !found {
+			return api.BadRequestError(fmt.Errorf("validator not found: %d", index))
+		}
+		pubkeys = append(pubkeys, share.ValidatorPubKey)
+	}
+
+	for _, req := range request.PubKeys {
+		var pubkey spectypes.ValidatorPK
+		if len(req) != len(pubkey) {
+			return api.BadRequestError(errors.New("invalid pubkey length"))
+		}
+		copy(pubkey[:], req)
+		pubkeys = append(pubkeys, pubkey)
+	}
+
+	var results []*dutytracer.ValidatorDutyTrace
+
+	for s := request.From; s <= request.To; s++ {
+		slot := phase0.Slot(s)
+		for _, r := range request.Roles {
+			role := spectypes.BeaconRole(r)
+			for _, pubkey := range pubkeys {
+				if isSyncCommitteeRole(role) {
+					committeeID, index, err := e.TraceStore.GetCommitteeID(slot, pubkey)
+					if err != nil {
+						return api.Error(fmt.Errorf("error getting committee ID: %w", err))
+					}
+					duty, err := e.TraceStore.GetCommitteeDuty(slot, committeeID, role)
+					if err != nil {
+						if errors.Is(err, dutytracer.ErrNotFound) {
+							// we might not have a duty for this role, so we skip it
+							continue
+						}
+						return api.Error(fmt.Errorf("error getting committee duty: %w", err))
+					}
+
+					validatorDuty := &dutytracer.ValidatorDutyTrace{
+						CommitteeID: committeeID,
+						ValidatorDutyTrace: model.ValidatorDutyTrace{
+							ConsensusTrace: duty.ConsensusTrace,
+							Slot:           duty.Slot,
+							Validator:      index,
+							Role:           role,
+						},
+					}
+
+					results = append(results, validatorDuty)
+
+					continue
+				}
+
+				duty, err := e.TraceStore.GetValidatorDuty(role, slot, pubkey)
+				if err != nil {
+					return api.Error(fmt.Errorf("error getting validator duties: %w", err))
+				}
+				results = append(results, duty)
+			}
+		}
+	}
+
+	return api.Render(w, r, toValidatorTraceResponse(results))
+}
+
+func isSyncCommitteeRole(role spectypes.BeaconRole) bool {
+	return role == spectypes.BNRoleSyncCommittee || role == spectypes.BNRoleAttester
+}
+
+var zeroCommitteeID spectypes.CommitteeID
```

Comment:
this variable is only used within `toValidatorTraceResponse` function, so it could be local to that function